README
This is "ptkftp", a Perl/Tk GUI interface to Perl's Net::FTP and Net::SFTP 
modules.  it requires either or both of them installed to be of much use.  
Without them all you can do is copy files between directories on your local 
machine. 

I wrote "ptkftp.pl" because I could not find a GUI program which worked with 
both FTP and SFTP/SSH2 for me.  It also takes full advantage of my 
Net::xFTP module.

Please note that this program is copyrighted by Jim Turner, (c) 2002 - 2008 
 and is released under the same license that Perl itself is.

This software is released in the 
hope that it will be useful, however the nether the author, the hosts of 
any websites from which the program may be obtained, nor anyone else for that 
matter is liable for any damages that may be caused by this program!  Use 
at your own risk, you have been warned!

To install, you must have Perl, Tk, and Net::xFTP modules installed.  Copy 
"JBrowseEntry.pl" and "JDialog.pl" to your Tk subdirectory where your Perl 
libraries are kept (or better yet - install them from CPAN).  
Then copy "JCutCopyPaste" to the same directory the "Tk" subdirectory 
resides, then copy "ptkftp.pl" to somewhere in your path. 

Invoke as:

ptkftp.pl &

feedback welcome at:  turnerjw784@yahoo.com

Enjoy the ride,

Jim Turner


Changes
Revision history for "ptkftp".

0.1   -Not released-
0.2   Wed Aug 14, 2002
	- original version released;
0.3   -Not released-
0.4   -Not released-
0.5	  Fri Aug 16, 2002
	- Made dropdown list selections to be saved on exit and reloaded when restarted.
	- Got working on Windows.
	- Got it to compile and build with Perl2Exe.
0.51  Sat Aug 17, 2002
	- Added "require" for my "SetPalette.pl" file.  Makes foreground white when 
	  background palette (-c="color") is set really dark.
0.52  -Not released-
0.53  Added checkbox to save password.  Password is encrypted before being 
	  saved to a text file called ".ptkftp.users", and is associated with a 
	  server and user-name.  This option is only available if the following 
	  modules are installed:  "Crypt::CBC", and either "Crypt::DES" or 
	  "Crypt::Blowfish_PP".
	- Added "-w" (width) option to specify fixed character width for the 
	  text and listboxes.
	- Now saves last user-name and directory under which a given host was 
	  logged in to.
0.54  Tue Aug 20, 2002
	- Remembered passwords now change when user is changed in dropdown list.
	- Fixed bug where some passwords were not stored.
	- Last directory accessed now saved with user instead of host (remembers 
	  which directory each user was last in on a given host.
0.55  -Not released-
0.56  -Not released-
0.57	 Sat Aug 24, 2002
	- Added "Edit, View, and Diff" options.  Selected remote file is auto-
	  matically downloaded to a temp file, and viewed, edited, or compared 
	  there.  If editing, it is automatically uploaded back when done.
	- Added Server-pinging option to keep connection open while idle.  Server 
	  is "ping"ed after 2 minutes, then at random intervals between 1 and 3 
	  minutes thereafter.  Pinging occurrs automatically if a "remote" file 
	  is opened for editing.
	- Added "file rename" option and dialog.
	- Added user-specified encryption module (-crypt=...) option.  Previously, 
	  "DES" was tried, then failing that, "Blowfish_PP".  The specified option 
	  is saved on exit.  Encryption is only used to encrypt "remembered" 
	  passwords when storing them to the ".ptkftp.users" file to prevent them 
	  from being viewable by anyone.  No encryption is necessary nor any 
	  encryption module needed to run the program if one does not wish the 
	  program to "remember" passwords.
0.59	 Sat Aug 31, 2002
	- Removed "keepopen" feature if running under Windows.
	- Added wheel-mouse support to scrollbars.
0.60	 Thu Feb 20, 2003
	- Fixed bug that messed up host-name in dropdown after disconnect and 
	  attempting to reconnect.
0.61  Unknown
	- Fixed bug where 2 sets of "." and ".." appeared in right-hand file list.
0.62  Thu June 2, 2005
	- Changed width to maxwidth in argument list for JBrowseEntry.
	- Improved selected directory history list and limited it to default of
	  last 20 directories selected.  Overrideable with "-h #".
0.80	- Completely redid to use Net::xFTP for all connections.  Also changed 
	  the file listings to use Tk::MListbox instead of Tk::Listbox so that 
	  file owner, group, permissions, and size could be displayed in 
	  columns along with the name.  One can also now see which files are 
	  directories.  Ascii/Binary radio buttons now greyed out except when 
	  using Net::FTP.
0.84 - Fixed some minor bugs.
0.86 - Fixed file listing bug in Windows - was trying to use "ls" command, 
	  which doesn't exist in Windows, now emulates "ls -l" without user/
	  group data.
0.90 - Fixed SFTP "input must be 8 bytes long at /usr/local/lib/perl/5.8.7/Crypt/DES.pm"
	  bug (actually didn't "fix", just worked around) by wrapping calls to 
	  sftp->new, sftp->put, and sftp->chmod along with their arguments in 
	  an eval string and evalling them?!?!?!  Must be something to do with 
	  Tk's en/decoding.  Placing the variable arguments as constants (in 
	  the eval string), seems to work.  Other functions don't seem to have 
	  problems.  Only had this happen on servers for one client.
0.91	- Fixed "Show All" - was broke somewhere in recent versions.
0.92	- Modified "Connect" button to say "connecting" and be disabled while 
	  connecting to a remote host.
1.00	- Added support for Net::SSH2 for doing Secure FTP (in addition to 
	  existing Net::SFTP support - by updating Net::xFTP to v. 0.20)
	  Created Windows binary and self-extracting setup exe.  Made compatable 
	  with Windows.


JCutCopyPaste.pl
#WRITTEN BY JIM TURNER (JWT) DUE TO COMMONALITY.
#
#PROVIDES THE 3 SUBROUTINES NEEDED TO DO CUT, COPY, AND PASTE OPERATIONS
#WITH THE CLIPBOARD FROM PERL/TK PROGRAMS.  THE ROUTINES ARE:  "doCut()",
#"doCopy()", "doPaste('clipboard')" (where 'clipboard' is 'PRIMARY', 'CLIPFILE'
#-if CLIPBOARD_FID is set, 
#'CLIPBOARD', 'SECONDARY', or other Unix clipboard; DEFAULT='CLIPBOARD'), 
#and "textfocusin()".
#
#usage:  require 'JCutCopyPaste.pl';
#...
#When you define a TEXT or ENTRY widget or any derivitive, follow it with 
#the binding:
#
#	$myText->bind('<FocusIn>' => [\&textfocusin]);
#
#The "do?????" routines can be called as callbacks from buttons, etc.
#
#THE PURPOSE OF TEXTFOCUSIN AND THE REQUIRED BINDING, ALLOWS CUTCOPYPASTE TO
#KEEP TRACK OF WHICH WIDGET HAS THE FOCUS ($ACTIVEWIDGET) FOR DOING THE
#CUTTING AND PASTING.  WHEN A TEXT OR ENTRY WIDGET RECEIVES THE FOCUS, 
#TEXTFOCUSIN SETS THE $ACTIVEWIDGET VARIABLE TO POINT TO THAT WIDGET.

sub doPaste
{
	my ($pasteboard) = shift;

	if (defined($activewidget))
	{
		my ($clipboard) = '';

		eval
		{
			$activewidget->markSet('selstartmk','insert');
			$activewidget->markGravity('selstartmk','left');
			$activewidget->markSet('selendmk','insert');
			$activewidget->markGravity('selendmk','right');
		};
		if ($pasteboard !~ /\S/ || $pasteboard eq 'PRIMARY')
		{
			eval {$clipboard = $MainWin->SelectionGet(-selection => 'PRIMARY');};
			$pasteboard = 'CLIPBOARD';
		}
		eval
		{
			if ($clipboard !~ /\S/ || $pasteboard eq 'CLIPFILE')
			{
				if (defined($ENV{CLIPBOARD_FID}))
				{
					if (open(CLIPBRD,"<$ENV{CLIPBOARD_FID}"))
					{
						$clipboard = join('',<CLIPBRD>);
						close CLIPBRD;
						#$activewidget->insert('insert',$clipboard);
						if (open(CLIPBRD,">$ENV{CLIPBOARD_FID}"))
						{
							binmode CLIPBRD;
							print CLIPBRD $clipboard;
							close CLIPBRD;
							if ($clipboard =~ /\S/)
							{
								$MainWin->SelectionOwn(-selection => 'CLIPBOARD');
								$MainWin->clipboardClear;
								$MainWin->clipboardAppend('--',$clipboard);
							}
						}
					}
				}
			}
		};
		eval
		{
			if ($clipboard !~ /\S/)
			{
				$clipboard = $MainWin->SelectionGet(-selection => $pasteboard);
			}
		};
		if ($clipboard =~ /\S/)
		{
			eval
			{
				$activewidget->insert('insert',$clipboard);
				#if ($activewidget->index('sel.first') =~ /\./)  #TEXT WIDGET
				{
					$activewidget->tagAdd('sel', 'selstartmk', 'selendmk');
				}
			};
		}
	}
}

sub doCopy
{
	if (defined($activewidget))
	{
		eval
		{
			$MainWin->SelectionOwn(-selection => 'CLIPBOARD');
			$MainWin->clipboardClear;
			if ($activewidget->index('sel.first') =~ /\./)  #TEXT WIDGET
			{
				$clipboard = $activewidget->get('sel.first','sel.last');
			}
			else  #ENTRY WIDGET
			{
				$clipboard = $activewidget->get;
				$clipboard = substr($clipboard,$activewidget->index('sel.first'),
					($activewidget->index('sel.last')
					- $activewidget->index('sel.first')));
			}
			$MainWin->clipboardAppend('--',$clipboard);
			if (defined($ENV{CLIPBOARD_FID}))
			{
				if (open(CLIPBRD,">$ENV{CLIPBOARD_FID}"))
				{
					binmode CLIPBRD;
					print CLIPBRD $clipboard;
					close CLIPBRD;
				}
			};
		}
	}
}

sub doCut
{
	if (defined($activewidget))
	{
		eval
		{
			$MainWin->SelectionOwn(-selection => 'CLIPBOARD');
			$MainWin->clipboardClear;
			if ($activewidget->index('sel.first') =~ /\./)  #TEXT WIDGET
			{
				$clipboard = $activewidget->get('sel.first','sel.last');
			}
			else  #ENTRY WIDGET
			{
				$clipboard = $activewidget->get;
				$clipboard = substr($clipboard,$activewidget->index('sel.first'),
					($activewidget->index('sel.last')
					- $activewidget->index('sel.first')));
			}
			$MainWin->clipboardAppend('--',$clipboard);
			if (defined($ENV{CLIPBOARD_FID}))
			{
				if (open(CLIPBRD,">$ENV{CLIPBOARD_FID}"))
				{
					binmode CLIPBRD;
					print CLIPBRD $clipboard;
					close CLIPBRD;
				}
			};
			$activewidget->delete('sel.first','sel.last');
		};
	}
}

sub textfocusin
{
	$activewidget = shift;
}

sub getactive
{
	return $activewidget;
}

1;



BindMouseWheel.pl
#!/usr/local/bin/perl -w

sub BindMouseWheel
{

	my($w) = @_;

	if ($^O eq 'MSWin32')
	{
		$w->bind('<MouseWheel>' =>
		[ sub { $_[0]->yview('scroll', -($_[1] / 120) * 1, 'units') },
				Ev('D') ]
		);
	}
	else
	{

		# Support for mousewheels on Linux commonly comes through
		# mapping the wheel to buttons 4 and 5.  If you have a
		# mousewheel ensure that the mouse protocol is set to
		# "IMPS/2" in your /etc/X11/XF86Config (or XF86Config-4)
		# file:
		#
		# Section "InputDevice"
		#     Identifier  "Mouse0"
		#     Driver      "mouse"
		#     Option      "Device" "/dev/mouse"
		#     Option      "Protocol" "IMPS/2"
		#     Option      "Emulate3Buttons" "off"
		#     Option      "ZAxisMapping" "4 5"
		# EndSection

		$MainWin->bind('<Alt-Left>' => sub
			{ 
				$_[0]->xview('scroll', -1, 'units');
				Tk->break;
			}
		);
		$MainWin->bind('<Alt-Right>' => sub
			{ 
				$_[0]->xview('scroll', +1, 'units');
				Tk->break;
			}
		);
		$MainWin->bind('<Alt-Button-4>' => sub
			{ 
					$_[0]->xview('scroll', -1, 'units');
					#$_[0]->yview('scroll', +1, 'units');
					Tk->break;
			}
		);
		$MainWin->bind('<Alt-Button-5>' => sub
			{ 
					$_[0]->xview('scroll', +1, 'units');
					#$_[0]->yview('scroll', -1, 'units');
					Tk->break;
			}
		);
		$w->bind('<Shift-Button-4>' => sub
		{
			$_[0]->yview('scroll', -1, 'pages') unless $Tk::strictMotif;
			Tk->break;
		}
		);

		$w->bind('<Shift-Button-5>' => sub
		{
			$_[0]->yview('scroll', +1, 'pages') unless $Tk::strictMotif;
			Tk->break;
		}
		);
		$w->bind('<Button-4>' => sub
		{
			$_[0]->yview('scroll', -1, 'units') unless $Tk::strictMotif;
		}
		);

		$w->bind('<Button-5>' => sub
		{
			$_[0]->yview('scroll', +1, 'units') unless $Tk::strictMotif;
		}
		);
	}

} # end BindMouseWheel

1


Tk/JDialog.pm
# JDialog - a translation of `tk_dialog' from Tcl/Tk to TkPerl (based on
# John Stoffel's idea).
#
# Modified 2/13/97 by Jim Turner of Computer Sciences Corporation to
# add underline character (alt-key) activation of buttons, fix bug in the
# bindings for <Return> key where default button always activated even if
# another button had the keyboard focus.  Now, the default button starts
# with the input focus!!!
#
# Jim Turner also added the "escape_button" option on 2/14/97 to allow
# programmer to specify a button to invoke if user presses the <Escape> key!
# Jim Turner also added the "images" option on 2/14/97 to allow programmer
# to specify gifs in leu of text for the buttons.
#
# Jim Turner also removed the "wraplength" option on 2/19/97 to allow
# longer label strings (>3") to not be broken.  User can specify -wraplength!
# Stephen O. Lidie, Lehigh University Computing Center.  94/12/27
# lusol@Lehigh.EDU
#
# 04/22/97 Jim Turner fixed bug where screen completely locks up if the calling
# script invokes a Motif app (ie. xv or another Perl/Tk app) shortly after    
# calling this dialog box.  Did not seem to adversely effect keyboard focus.
# fixed by commenting out 1 line of code (&$old_focus);
#
# This is an OO implementation of `tk_dialog'.  First, create all your Dialog
# objects during program initialization.  When it's time to use a dialog, 
# invoke the `show' method on a dialog object; the method then displays the 
# dialog, waits for a button to be invoked, and returns the text label of the 
# selected button.
#
# A Dialog object essentially consists of two subwidgets: a Label widget for
# the bitmap and a Label wigdet for the text of the dialog.  If required, you 
# can invoke the `configure' method to change any characteristic of these 
# subwidgets.
#
# Because a Dialog object is a Toplevel widget all the 'composite' base class
# methods are available to you.
#
# Advertised widgets:  bitmap, message.
#
# 1) Call the constructor to create the dialog object, which in turn returns 
#    a blessed reference to the new composite widget:
#
#    require Tk::JDialog;
#
#    $DialogRef = $mw->JDialog(
#        -title          => $title,
#        -text           => $text,
#        -bitmap         => $bitmap,
#        -default_button => $default_button,
#        -escape_button  => $escape_button,
#        -buttons        => [@button_labels],
#	 -images         => ['gif-file1','gif-file2'...]
#    );
#
#       mw             - a window reference, usually the result of a
#                        MainWindow->new call.
#       title          - Title to display in the dialog's decorative frame.
#                        NOTE: Jim Turner added optional tilde to "title"
#                        to specify a character to be the "underline" char.
#       text           - Message to display in the dialog widget.
#       bitmap         - Bitmap to display in the dialog.
#       default_button - Text label of the button that is to display the
#                        default ring (''signifies no default button).
#       escape_button  - Text label of the button that is to be invoked 
#                        when the user presses the <Esc> key.
#       button_labels  - A reference to a list of one or more strings to
#                        display in buttons across the bottom of the dialog.
#       images         - Specify the path and fid for an image or '' if a
#                        corresponding button is to use text.  NOTE: button
#                        will use text if the image file is not found and
#                        that the "-button_labels" option MUST ALWAYS be 
#                        specified anyway!
#       noballoons     - if true, do not display the "button_labels" value 
#                        in a balloon if images are used.
#
# 2) Invoke the `show' method on a dialog object:
#
#    $button_label = $DialogRef->Show;
#
#       This returns the text label of the selected button.
#
#    (Note:  you can request a global grab by passing the string "-global"
#    to the `show' method.)

package Tk::JDialog;
use Carp;
use strict qw(vars);
our $useBalloon;
use  Tk::Balloon; $useBalloon = 1;
require Tk::Toplevel;
@Tk::JDialog::ISA = qw(Tk::Toplevel);

Tk::Widget->Construct('JDialog');

sub Populate
{

    # Dialog object constructor.  Uses `new' method from base class
    # to create object container then creates the dialog toplevel.

    my($cw, $args) = @_;

    $cw->SUPER::Populate($args);

    my ($w_bitmap,$w_but,$pad1,$pad2,$underlinepos,$mychar,$blshow,$i);
    my ($btnopt,$undopt,$balloon);

    my $buttons = delete $args->{'-buttons'};
    my $images = delete $args->{'-images'};
    $buttons = ['OK'] unless (defined $buttons);
    my $default_button = delete $args->{-default_button};
    my $escape_button = delete $args->{-escape_button};
    my $noballoons = delete $args->{-noballoons};
    $useBalloon = 0  if ($noballoons);
    $default_button =  $buttons->[0] unless (defined $default_button);

    # Create the Toplevel window and divide it into top and bottom parts.

    $cw->{'selected_button'} = '';
    my (@pl) = (-side => 'top', -fill => 'both');
    ($pad1, $pad2) =
        ([-padx => '3m', -pady => '3m'], [-padx => '3m', -pady => '2m']);

    $cw->withdraw;
    $cw->iconname('JDialog');
    $cw->protocol('WM_DELETE_WINDOW' => sub {});
#?????????????????    $cw->transient($cw->toplevel)  unless ($^O =~ /Win/i);

    my $w_top = $cw->Frame(Name => 'top',-relief => 'raised', -borderwidth => 1);
    my $w_bot = $cw->Frame(Name => 'bot',-relief => 'raised', -borderwidth => 1);
    $w_top->pack(@pl);
    $w_bot->pack(@pl);

    # Fill the top part with the bitmap and message.

    @pl = (-side => 'left');

    $w_bitmap = $w_top->Label(Name => 'bitmap');
    $w_bitmap->pack(@pl, @$pad1);
    my $w_msg = $w_top->Label(
	#-wraplength => '3i',    --!!! Removed 2/19 by Jim Turner
	-justify    => 'left' );

    $w_msg->pack(-side => 'right', -expand => 1, -fill => 'both', @$pad1);

    # Create a row of buttons at the bottom of the dialog.

    my($w_default_button, $bl) = (undef, '');
    $cw->{'default_button'} = undef;
    $cw->{'escape_button'} = undef;
    $i = 0;
    foreach $bl (@$buttons) {

	$blshow = $bl;
	$underlinepos = ($blshow =~ s/^(.*)~/$1/) ? length($1): undef;
	if (defined($$images[$i]) && $$images[$i] gt ' ' && -e $$images[$i])
	{
		$cw->Photo($blshow, -file => $$images[$i]);
		$btnopt = '-image';
	}
	else
	{
		$btnopt = '-text';
	}
	if (defined($underlinepos))
	{
		$mychar = substr($blshow,$underlinepos,1);
		$w_but = $w_bot->Button(
		    $btnopt => $blshow,
		    -underline => $underlinepos,
		    -command => [
			sub {
			    $_[0]->{'selected_button'} = $_[1];
			}, $cw, $bl,
		    ]
		);
		$cw->bind("<Alt-\l$mychar>", [$w_but => "Invoke"]);
		$cw->bind("<Key-\l$mychar>", [$w_but => "Invoke"]);
	}
	else
	{
		$w_but = $w_bot->Button(
		    $btnopt => $blshow,
		    -command => [
			sub {
			    $_[0]->{'selected_button'} = $_[1];
			}, $cw, $bl,
		    ]
		);
	}
	if ($useBalloon && $btnopt eq '-image')
	{
		$balloon = $cw->Balloon();
		$balloon->attach($w_but, -state => 'balloon', -balloonmsg => $blshow);
	}
        if ($bl eq $default_button) {
            $w_default_button = $w_bot->Frame(
                -relief      => 'sunken',
                -borderwidth => 1
            );
            $w_but->raise($w_default_button);
            $w_default_button->pack(@pl, -expand => 1, @$pad2);
            $w_but->pack(-in => $w_default_button, -padx => '2m',
                         -pady => '2m');
    	    $cw->{'default_button'} = $w_but;
goto JWT_SKIP1;
            $cw->bind(
                '<Return>' => [
                    sub {
                        $_[1]->flash; 
                        $_[2]->{'selected_button'} = $_[3];
                    }, $w_but, $cw, $bl,
                ]
            );
JWT_SKIP1:
        } else {
            $w_but->pack(@pl, -expand => 1, @$pad2);
    	    $cw->{'default_button'} = $w_but  unless(defined($cw->{'default_button'}));
        }
	if ($bl eq $escape_button)
	{
            $cw->{'escape_button'} = $w_but;
            $cw->bind('<Escape>' => [$w_but => "Invoke"]);
	}
        ++$i;
    } # forend all buttons

    $cw->Advertise(message => $w_msg);
    $cw->Advertise(bitmap  => $w_bitmap );
    #!!!$cw->{'default_button'} = $w_default_button;

	if ($^O =~ /Win/i)
	{
		$cw->ConfigSpecs(
				-image      => ['bitmap',undef,undef,undef],
				-bitmap     => ['bitmap',undef,undef,undef],
				-fg         => ['ADVERTISED','foreground','Foreground','black'],
				-foreground => ['ADVERTISED','foreground','Foreground','black'],
				-bg         => ['DESCENDANTS','background','Background',undef],
				-background => ['DESCENDANTS','background','Background',undef],
		DEFAULT     => ['message',undef,undef,undef]
		);
	}
	else
	{
		$cw->ConfigSpecs(
				-image      => ['bitmap',undef,undef,undef],
				-bitmap     => ['bitmap',undef,undef,undef],
				-fg         => ['ADVERTISED','foreground','Foreground','black'],
				-foreground => ['ADVERTISED','foreground','Foreground','black'],
				-bg         => ['DESCENDANTS','background','Background',undef],
				-background => ['DESCENDANTS','background','Background',undef],
		# JWT for TNT!  -font       => ['message','font','Font','-*-Times-Medium-R-Normal-*-180-*-*-*-*-*-*'],
				-font       => ['message','font','Font','-adobe-helvetica-bold-r-normal--17-120-100-100-p-92-iso8859-1'],
		DEFAULT     => ['message',undef,undef,undef]
		);
	}
} # end Dialog constructor

sub Show {

    # Dialog object public method - display the dialog.

    my ($cw, $grab_type) = @_;

    croak "Dialog:  `show' method requires at least 1 argument"
        if scalar @_ < 1 ;

    #!!! my $old_focus = $cw->focusSave;  # don't need (Jim Turner) after fixing BUG!
    my $old_grab  = $cw->grabSave;

    # Update all geometry information, center the dialog in the display
    # and deiconify it

    $cw->Popup(); 

    # set a grab and claim the focus.

		 unless ($ENV{DESKTOP_SESSION} =~ /kde/)
		 {
	    if (defined $grab_type && length $grab_type) {
	        $cw->grab($grab_type)  if ($grab_type !~ /no/i);  #JWT: ADDED 20010517 TO ALLOW NON-GRABBING!
	    } else {
	        $cw->grab;
	    }
		 }
############## $cw->waitVisibility;  #SEEMS TO HANG ON NEWER TK'S.
	$cw->update;
    if (defined $cw->{'default_button'}) 
     {
      $cw->{'default_button'}->focus;
     } 
    else
     {
      $cw->focus;
     }

    # Wait for the user to respond, restore the focus and grab, withdraw
    # the dialog and return the label of the selected button.

    $cw->waitVariable(\$cw->{'selected_button'});
    $cw->grabRelease;
    $cw->withdraw;
    #!!! &$old_focus;   #FIXED BUG CAUSING COMPLETE SCREEN LOCKUP IF ANOTHER
			#MOTIF APP (WINDOW) IS POPPED UP SHORTLY AFTERWARDS!
    &$old_grab;
    return $cw->{'selected_button'};

} # end Dialog show method

1;
